File: crates/stateset-embedded/src/quality.rs
//! Quality Control operations
//!
//! Comprehensive quality management system supporting:
//! - Inspections (receiving, in-process, final, random)
//! - Non-conformance reports (NCRs)
//! - Quality holds on inventory
//! - Defect code management
//!
//! # Example
//!
//! ```rust,ignore
//! use stateset_embedded::{Commerce, CreateInspection, InspectionType};
//! use uuid::Uuid;
//!
//! let commerce = Commerce::new("./store.db")?;
//!
//! // Create an inspection for received goods
//! let inspection = commerce.quality().create_inspection(CreateInspection {
//!     inspection_type: InspectionType::Receiving,
//!     reference_type: "purchase_order".into(),
//!     reference_id: Uuid::new_v4(),
//!     ..Default::default()
//! })?;
//!
//! println!("Created inspection #{}", inspection.inspection_number);
//! # Ok::<(), stateset_embedded::CommerceError>(())
//! ```

use stateset_core::{
    CreateDefectCode, CreateInspection, CreateNonConformance, CreateQualityHold,
    DefectCode, Inspection, InspectionFilter, InspectionItem, InspectionStatus,
    NonConformance, NonConformanceFilter, NcrStatus, QualityHold, QualityHoldFilter,
    QualityRepository, RecordInspectionResult, ReleaseQualityHold, Result,
    UpdateInspection, UpdateNonConformance,
};
use stateset_db::sqlite::SqliteQualityRepository;
use uuid::Uuid;

/// Quality control management interface.
pub struct Quality {
    repo: SqliteQualityRepository,
}

impl Quality {
    pub(crate) fn new(repo: SqliteQualityRepository) -> Self {
        Self { repo }
    }

    // ========================================================================
    // Inspections
    // ========================================================================

    /// Create a new inspection.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, CreateInspection, InspectionType};
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let inspection = commerce.quality().create_inspection(CreateInspection {
    ///     inspection_type: InspectionType::Receiving,
    ///     reference_type: "purchase_order".into(),
    ///     reference_id: Uuid::new_v4(),
    ///     ..Default::default()
    /// })?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn create_inspection(&self, input: CreateInspection) -> Result<Inspection> {
        self.repo.create_inspection(input)
    }

    /// Get an inspection by ID.
    pub fn get_inspection(&self, id: Uuid) -> Result<Option<Inspection>> {
        self.repo.get_inspection(id)
    }

    /// Get an inspection by its number.
    pub fn get_inspection_by_number(&self, number: &str) -> Result<Option<Inspection>> {
        self.repo.get_inspection_by_number(number)
    }

    /// List inspections with optional filtering.
    pub fn list_inspections(&self, filter: InspectionFilter) -> Result<Vec<Inspection>> {
        self.repo.list_inspections(filter)
    }

    /// Update an inspection.
    pub fn update_inspection(&self, id: Uuid, input: UpdateInspection) -> Result<Inspection> {
        self.repo.update_inspection(id, input)
    }

    /// Start an inspection (set status to InProgress).
    pub fn start_inspection(&self, id: Uuid) -> Result<Inspection> {
        self.repo.start_inspection(id)
    }

    /// Record inspection results for an item.
    pub fn record_inspection_result(&self, input: RecordInspectionResult) -> Result<InspectionItem> {
        self.repo.record_inspection_result(input)
    }

    /// Get inspection items for an inspection.
    pub fn get_inspection_items(&self, inspection_id: Uuid) -> Result<Vec<InspectionItem>> {
        self.repo.get_inspection_items(inspection_id)
    }

    /// Complete an inspection (mark as Passed or Failed based on results).
    pub fn complete_inspection(&self, id: Uuid) -> Result<Inspection> {
        self.repo.complete_inspection(id)
    }

    /// Delete an inspection (only if Pending).
    pub fn delete_inspection(&self, id: Uuid) -> Result<()> {
        self.repo.delete_inspection(id)
    }

    /// Count inspections matching filter.
    pub fn count_inspections(&self, filter: InspectionFilter) -> Result<u64> {
        self.repo.count_inspections(filter)
    }

    // ========================================================================
    // Non-Conformance Reports
    // ========================================================================

    /// Create a non-conformance report (NCR).
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, CreateNonConformance, NonConformanceSource, Severity};
    /// use rust_decimal_macros::dec;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let ncr = commerce.quality().create_ncr(CreateNonConformance {
    ///     source: NonConformanceSource::Inspection,
    ///     severity: Severity::Major,
    ///     sku: "SKU-001".into(),
    ///     quantity_affected: dec!(10),
    ///     description: "Parts do not meet specification".into(),
    ///     ..Default::default()
    /// })?;
    ///
    /// println!("Created NCR #{}", ncr.ncr_number);
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn create_ncr(&self, input: CreateNonConformance) -> Result<NonConformance> {
        self.repo.create_ncr(input)
    }

    /// Get a non-conformance report by ID.
    pub fn get_ncr(&self, id: Uuid) -> Result<Option<NonConformance>> {
        self.repo.get_ncr(id)
    }

    /// Get a non-conformance report by its number.
    pub fn get_ncr_by_number(&self, number: &str) -> Result<Option<NonConformance>> {
        self.repo.get_ncr_by_number(number)
    }

    /// List non-conformance reports with optional filtering.
    pub fn list_ncrs(&self, filter: NonConformanceFilter) -> Result<Vec<NonConformance>> {
        self.repo.list_ncrs(filter)
    }

    /// Update a non-conformance report.
    ///
    /// Use this to set root cause, corrective action, disposition, etc.
    pub fn update_ncr(&self, id: Uuid, input: UpdateNonConformance) -> Result<NonConformance> {
        self.repo.update_ncr(id, input)
    }

    /// Close an NCR.
    pub fn close_ncr(&self, id: Uuid) -> Result<NonConformance> {
        self.repo.close_ncr(id)
    }

    /// Cancel an NCR.
    pub fn cancel_ncr(&self, id: Uuid) -> Result<NonConformance> {
        self.repo.cancel_ncr(id)
    }

    /// Count NCRs matching filter.
    pub fn count_ncrs(&self, filter: NonConformanceFilter) -> Result<u64> {
        self.repo.count_ncrs(filter)
    }

    // ========================================================================
    // Quality Holds
    // ========================================================================

    /// Create a quality hold on inventory.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, CreateQualityHold, HoldType};
    /// use rust_decimal_macros::dec;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let hold = commerce.quality().create_hold(CreateQualityHold {
    ///     sku: "SKU-001".into(),
    ///     lot_number: Some("LOT-2025-001".into()),
    ///     quantity_held: dec!(50),
    ///     reason: "Pending quality inspection".into(),
    ///     hold_type: HoldType::QualityInspection,
    ///     placed_by: Some("QA Team".into()),
    ///     ..Default::default()
    /// })?;
    ///
    /// println!("Hold placed on {} units", hold.quantity_held);
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn create_hold(&self, input: CreateQualityHold) -> Result<QualityHold> {
        self.repo.create_hold(input)
    }

    /// Get a quality hold by ID.
    pub fn get_hold(&self, id: Uuid) -> Result<Option<QualityHold>> {
        self.repo.get_hold(id)
    }

    /// List quality holds with optional filtering.
    pub fn list_holds(&self, filter: QualityHoldFilter) -> Result<Vec<QualityHold>> {
        self.repo.list_holds(filter)
    }

    /// Release a quality hold.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, ReleaseQualityHold};
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// commerce.quality().release_hold(Uuid::new_v4(), ReleaseQualityHold {
    ///     released_by: "QA Manager".into(),
    ///     notes: Some("Inspection passed".into()),
    /// })?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn release_hold(&self, id: Uuid, input: ReleaseQualityHold) -> Result<QualityHold> {
        self.repo.release_hold(id, input)
    }

    /// Get active holds for a SKU.
    pub fn get_active_holds_for_sku(&self, sku: &str) -> Result<Vec<QualityHold>> {
        self.repo.get_active_holds_for_sku(sku)
    }

    /// Get active holds for a lot.
    pub fn get_active_holds_for_lot(&self, lot_number: &str) -> Result<Vec<QualityHold>> {
        self.repo.get_active_holds_for_lot(lot_number)
    }

    /// Count active holds.
    pub fn count_active_holds(&self) -> Result<u64> {
        self.repo.count_active_holds()
    }

    // ========================================================================
    // Defect Codes
    // ========================================================================

    /// Create a defect code.
    pub fn create_defect_code(&self, input: CreateDefectCode) -> Result<DefectCode> {
        self.repo.create_defect_code(input)
    }

    /// Get a defect code by code.
    pub fn get_defect_code(&self, code: &str) -> Result<Option<DefectCode>> {
        self.repo.get_defect_code(code)
    }

    /// List all defect codes, optionally filtered by category.
    pub fn list_defect_codes(&self, category: Option<&str>) -> Result<Vec<DefectCode>> {
        self.repo.list_defect_codes(category)
    }

    /// Deactivate a defect code.
    pub fn deactivate_defect_code(&self, id: Uuid) -> Result<()> {
        self.repo.deactivate_defect_code(id)
    }

    // ========================================================================
    // Convenience Methods
    // ========================================================================

    /// Get all pending inspections.
    pub fn get_pending_inspections(&self) -> Result<Vec<Inspection>> {
        self.list_inspections(InspectionFilter {
            status: Some(InspectionStatus::Pending),
            ..Default::default()
        })
    }

    /// Get all open NCRs.
    pub fn get_open_ncrs(&self) -> Result<Vec<NonConformance>> {
        self.list_ncrs(NonConformanceFilter {
            status: Some(NcrStatus::Open),
            ..Default::default()
        })
    }

    /// Get all active holds.
    pub fn get_active_holds(&self) -> Result<Vec<QualityHold>> {
        self.list_holds(QualityHoldFilter {
            active_only: Some(true),
            ..Default::default()
        })
    }
}

File: cli/src/commands/quality.js
/**
 * Quality Commands Module
 */

function parsePositive(value, usage) {
  const parsed = Number.parseFloat(value);
  if (!Number.isFinite(parsed) || parsed <= 0) throw new Error(usage);
  return parsed;
}

function parseOptionalInt(value, usage) {
  if (!value) return undefined;
  const parsed = Number.parseInt(value, 10);
  if (!Number.isInteger(parsed)) throw new Error(usage);
  return parsed;
}

export async function execute(action, args, { commerce, output, jsonOutput }) {
  switch (action) {
    case 'inspections': {
      const inspections = await commerce.quality.listInspections();
      return formatInspections(inspections, { output, jsonOutput });
    }

    case 'inspection': {
      const inspectionId = args[0];
      if (!inspectionId) throw new Error('Usage: quality inspection <inspectionId>');
      const inspection = await commerce.quality.getInspection(inspectionId);
      if (!inspection) throw new Error(`Inspection not found: ${inspectionId}`);
      return formatInspection(inspection, { jsonOutput });
    }

    case 'create-inspection': {
      const [inspectionType, referenceType, referenceId, warehouseIdRaw, assignedTo, ...noteParts] =
        args;
      if (!inspectionType || !referenceType || !referenceId) {
        throw new Error(
          'Usage: quality create-inspection <inspectionType> <referenceType> <referenceId> [warehouseId] [assignedTo] [notes]',
        );
      }
      const inspection = await commerce.quality.createInspection({
        inspectionType,
        referenceType,
        referenceId,
        warehouseId: parseOptionalInt(warehouseIdRaw, 'warehouseId must be an integer'),
        assignedTo: assignedTo || undefined,
        notes: noteParts.join(' ') || undefined,
      });
      return { inspection, formatted: `Created inspection ${inspection.id}` };
    }

    case 'start-inspection': {
      const inspectionId = args[0];
      if (!inspectionId) throw new Error('Usage: quality start-inspection <inspectionId>');
      const inspection = await commerce.quality.startInspection(inspectionId);
      return { inspection, formatted: `Started inspection ${inspection.id}` };
    }

    case 'complete-inspection': {
      const inspectionId = args[0];
      if (!inspectionId) throw new Error('Usage: quality complete-inspection <inspectionId>');
      const inspection = await commerce.quality.completeInspection(inspectionId);
      return { inspection, formatted: `Completed inspection ${inspection.id}` };
    }

    case 'ncrs': {
      const ncrs = await commerce.quality.listNcrs();
      return formatNcrs(ncrs, { output, jsonOutput });
    }

    case 'ncr': {
      const ncrId = args[0];
      if (!ncrId) throw new Error('Usage: quality ncr <ncrId>');
      const ncr = await commerce.quality.getNcr(ncrId);
      if (!ncr) throw new Error(`NCR not found: ${ncrId}`);
      return formatNcr(ncr, { jsonOutput });
    }

    case 'create-ncr': {
      const [source, severity, sku, quantityAffectedRaw, description, lotNumber, locationIdRaw] =
        args;
      if (!source || !severity || !sku || !quantityAffectedRaw || !description) {
        throw new Error(
          'Usage: quality create-ncr <source> <severity> <sku> <quantityAffected> <description> [lotNumber] [locationId]',
        );
      }
      const ncr = await commerce.quality.createNcr({
        source,
        severity,
        sku,
        quantityAffected: parsePositive(
          quantityAffectedRaw,
          'Usage: quality create-ncr <source> <severity> <sku> <quantityAffected> <description> [lotNumber] [locationId]',
        ),
        description,
        lotNumber: lotNumber || undefined,
        locationId: parseOptionalInt(locationIdRaw, 'locationId must be an integer'),
      });
      return { ncr, formatted: `Created NCR ${ncr.id}` };
    }

    case 'close-ncr': {
      const ncrId = args[0];
      if (!ncrId) throw new Error('Usage: quality close-ncr <ncrId>');
      const ncr = await commerce.quality.closeNcr(ncrId);
      return { ncr, formatted: `Closed NCR ${ncr.id}` };
    }

    case 'holds': {
      const holds = await commerce.quality.listHolds();
      return formatHolds(holds, { output, jsonOutput });
    }

    case 'hold': {
      const holdId = args[0];
      if (!holdId) throw new Error('Usage: quality hold <holdId>');
      const hold = await commerce.quality.getHold(holdId);
      if (!hold) throw new Error(`Quality hold not found: ${holdId}`);
      return formatHold(hold, { jsonOutput });
    }

    case 'create-hold': {
      const [sku, lotNumber, quantityHeldRaw, reason, holdType, placedBy, locationIdRaw] = args;
      if (!sku || !quantityHeldRaw || !reason || !holdType) {
        throw new Error(
          'Usage: quality create-hold <sku> [lotNumber] <quantityHeld> <reason> <holdType> [placedBy] [locationId]',
        );
      }
      const hold = await commerce.quality.createHold({
        sku,
        lotNumber: lotNumber || undefined,
        quantityHeld: parsePositive(
          quantityHeldRaw,
          'Usage: quality create-hold <sku> [lotNumber] <quantityHeld> <reason> <holdType> [placedBy] [locationId]',
        ),
        reason,
        holdType,
        placedBy: placedBy || undefined,
        locationId: parseOptionalInt(locationIdRaw, 'locationId must be an integer'),
      });
      return { hold, formatted: `Created quality hold ${hold.id}` };
    }

    case 'release-hold': {
      const [holdId, releasedBy, ...noteParts] = args;
      if (!holdId || !releasedBy) {
        throw new Error('Usage: quality release-hold <holdId> <releasedBy> [notes]');
      }
      const hold = await commerce.quality.releaseHold(
        holdId,
        releasedBy,
        noteParts.join(' ') || undefined,
      );
      return { hold, formatted: `Released quality hold ${hold.id}` };
    }

    case 'active-holds': {
      const holds = await commerce.quality.getActiveHolds();
      return formatHolds(holds, { output, jsonOutput });
    }

    case 'count': {
      const count = await commerce.quality.countActiveHolds();
      return { count, formatted: `Active quality hold count: ${count}` };
    }

    default:
      throw new Error(
        `Unknown action: quality ${action}\n\n` +
          'Available actions:\n' +
          '  inspections                                                            List inspections\n' +
          '  inspection <inspectionId>                                              Get inspection\n' +
          '  create-inspection <inspectionType> <referenceType> <referenceId> [warehouseId] [assignedTo] [notes]\n' +
          '  start-inspection <inspectionId>                                        Start inspection\n' +
          '  complete-inspection <inspectionId>                                     Complete inspection\n' +
          '  ncrs                                                                   List NCRs\n' +
          '  ncr <ncrId>                                                            Get NCR\n' +
          '  create-ncr <source> <severity> <sku> <quantityAffected> <description> [lotNumber] [locationId]\n' +
          '  close-ncr <ncrId>                                                      Close NCR\n' +
          '  holds                                                                  List quality holds\n' +
          '  hold <holdId>                                                          Get quality hold\n' +
          '  create-hold <sku> [lotNumber] <quantityHeld> <reason> <holdType> [placedBy] [locationId]\n' +
          '  release-hold <holdId> <releasedBy> [notes]                             Release quality hold\n' +
          '  active-holds                                                           List active quality holds\n' +
          '  count                                                                  Count active quality holds',
      );
  }
}

function formatInspections(inspections, { output, jsonOutput }) {
  if (jsonOutput) return inspections;
  if (inspections.length === 0) return { formatted: 'No inspections found.' };
  const formatted = output.table(inspections, [
    { key: 'id', header: 'ID' },
    { key: 'inspectionType', header: 'Type' },
    { key: 'referenceType', header: 'Ref Type' },
    { key: 'referenceId', header: 'Ref ID' },
    { key: 'status', header: 'Status' },
    { key: 'assignedTo', header: 'Assigned To' },
  ]);
  return { inspections, formatted };
}

function formatInspection(inspection, { jsonOutput }) {
  if (jsonOutput) return inspection;
  return {
    inspection,
    formatted:
      `Inspection: ${inspection.id}\n` +
      `${'-'.repeat(34)}\n` +
      `Type:           ${inspection.inspectionType}\n` +
      `Reference:      ${inspection.referenceType}:${inspection.referenceId}\n` +
      `Warehouse:      ${inspection.warehouseId || 'N/A'}\n` +
      `Assigned:       ${inspection.assignedTo || 'N/A'}\n` +
      `Status:         ${inspection.status}`,
  };
}

function formatNcrs(ncrs, { output, jsonOutput }) {
  if (jsonOutput) return ncrs;
  if (ncrs.length === 0) return { formatted: 'No NCRs found.' };
  const formatted = output.table(ncrs, [
    { key: 'id', header: 'ID' },
    { key: 'source', header: 'Source' },
    { key: 'severity', header: 'Severity' },
    { key: 'sku', header: 'SKU' },
    { key: 'quantityAffected', header: 'Qty', align: 'right' },
    { key: 'status', header: 'Status' },
  ]);
  return { ncrs, formatted };
}

function formatNcr(ncr, { jsonOutput }) {
  if (jsonOutput) return ncr;
  return {
    ncr,
    formatted:
      `NCR: ${ncr.id}\n` +
      `${'-'.repeat(22)}\n` +
      `Source:         ${ncr.source}\n` +
      `Severity:       ${ncr.severity}\n` +
      `SKU:            ${ncr.sku}\n` +
      `Qty affected:   ${ncr.quantityAffected}\n` +
      `Status:         ${ncr.status}`,
  };
}

function formatHolds(holds, { output, jsonOutput }) {
  if (jsonOutput) return holds;
  if (holds.length === 0) return { formatted: 'No quality holds found.' };
  const formatted = output.table(holds, [
    { key: 'id', header: 'ID' },
    { key: 'sku', header: 'SKU' },
    { key: 'lotNumber', header: 'Lot' },
    { key: 'quantityHeld', header: 'Qty', align: 'right' },
    { key: 'holdType', header: 'Type' },
    { key: 'status', header: 'Status' },
  ]);
  return { holds, formatted };
}

function formatHold(hold, { jsonOutput }) {
  if (jsonOutput) return hold;
  return {
    hold,
    formatted:
      `Quality hold: ${hold.id}\n` +
      `${'-'.repeat(30)}\n` +
      `SKU:            ${hold.sku}\n` +
      `Lot:            ${hold.lotNumber || 'N/A'}\n` +
      `Quantity held:  ${hold.quantityHeld}\n` +
      `Type:           ${hold.holdType}\n` +
      `Status:         ${hold.status}\n` +
      `Reason:         ${hold.reason}`,
  };
}

export const metadata = {
  name: 'quality',
  aliases: ['qa', 'ncr'],
  description: 'Quality inspections, NCRs, and holds',
  actions: {
    inspections: { description: 'List inspections', args: [] },
    inspection: { description: 'Get inspection', args: ['<inspectionId>'] },
    'create-inspection': {
      description: 'Create inspection',
      args: [
        '<inspectionType>',
        '<referenceType>',
        '<referenceId>',
        '[warehouseId]',
        '[assignedTo]',
        '[notes]',
      ],
    },
    'start-inspection': { description: 'Start inspection', args: ['<inspectionId>'] },
    'complete-inspection': { description: 'Complete inspection', args: ['<inspectionId>'] },
    ncrs: { description: 'List NCRs', args: [] },
    ncr: { description: 'Get NCR', args: ['<ncrId>'] },
    'create-ncr': {
      description: 'Create NCR',
      args: [
        '<source>',
        '<severity>',
        '<sku>',
        '<quantityAffected>',
        '<description>',
        '[lotNumber]',
        '[locationId]',
      ],
    },
    'close-ncr': { description: 'Close NCR', args: ['<ncrId>'] },
    holds: { description: 'List quality holds', args: [] },
    hold: { description: 'Get quality hold', args: ['<holdId>'] },
    'create-hold': {
      description: 'Create quality hold',
      args: [
        '<sku>',
        '[lotNumber]',
        '<quantityHeld>',
        '<reason>',
        '<holdType>',
        '[placedBy]',
        '[locationId]',
      ],
    },
    'release-hold': {
      description: 'Release quality hold',
      args: ['<holdId>', '<releasedBy>', '[notes]'],
    },
    'active-holds': { description: 'List active quality holds', args: [] },
    count: { description: 'Count active quality holds', args: [] },
  },
};

export default { execute, metadata };
